File: week3/승진/BJ16987_계란으로 계란치기.java
import java.util.Scanner;

public class Main {
    static int N;
    static int[] S;
    static int[] W;
    static int max_cnt;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();
        S = new int[N];
        W = new int[N];
        max_cnt = 0;

        for (int i = 0; i < N; i++) {
            S[i] = scanner.nextInt();
            W[i] = scanner.nextInt();
        }

        backtrack(0, 0);
        System.out.println(max_cnt);
    }

    public static void backtrack(int cur, int cnt) {
        if (cur == N) {
            if (cnt > max_cnt) max_cnt = cnt;
            return;
        }

        if (S[cur] <= 0 || cnt == N - 1) {
            backtrack(cur + 1, cnt);
            return;
        }

        for (int i = 0; i < N; i++) {
            if (i == cur || S[i] <= 0) continue;

            // update S, W
            S[cur] -= W[i];
            S[i] -= W[cur];
            if (S[cur] <= 0) cnt++;
            if (S[i] <= 0) cnt++;

            backtrack(cur + 1, cnt);

            // restore S, W
            if (S[cur] <= 0) cnt--;
            if (S[i] <= 0) cnt--;
            S[cur] += W[i];
            S[i] += W[cur];
        }
    }
}

File: week3/승진/BJ16987_계란으로 계란치기.py
def backtrack(cur, cnt):
    if cur == N:
        global max_cnt
        if cnt > max_cnt: max_cnt = cnt
        return
    
    if S[cur] <= 0 or cnt == N - 1: 
        backtrack(cur + 1, cnt)
        return

    for i in range(N):
        if i == cur or S[i] <= 0:
            continue
        # update s, w
        S[cur] -= W[i]
        S[i] -= W[cur]
        if S[cur] <= 0: cnt += 1
        if S[i] <= 0: cnt += 1

        backtrack(cur + 1, cnt)
        # restore s, w
        if S[cur] <= 0: cnt -= 1
        if S[i] <= 0: cnt -= 1
        S[cur] += W[i]
        S[i] += W[cur]
        

        
N = int(input())
S = []
W = []

max_cnt = 0

for _ in range(N):
    s, w = map(int, input().split())
    S.append(s)
    W.append(w)

backtrack(0, 0)
print(max_cnt)